### DISCUSSION TITLE
Are Styles supposed to Initiate IBindings even when they don't match exactly?

### DISCUSSION THREAD

**Discussion by RationalFragile** [AvaloniaUI/Avalonia#7319] [Q&A]
Hi, I'm not sure if this is by design or I'm doing something wrong, but basically, if you have a style like this
```
<Style Selector="Button.Red">
    <Setter Property="Background" Value="{DynamicResource buttonBackground}"/>
</Style>
```
every single button calls the
```
InstancedBinding? IBinding.Initiate(IAvaloniaObject target,
      AvaloniaProperty targetProperty,
      object anchor,
      bool enableDataValidation);
```
method on the `IBinding`. (I know because I have my own custom DynamicResourceExtension...)

And so it creates a large number of `InstancedBinding`s: If I have 100 buttons, and 10 button classes, it will create 10x100 `InstancedBinding`s even if none of the buttons use any of those styles. This is because when initializing the `IStyleable`, the `Selector` matches the button with a `SelectorMatchResult.Sometimes`.  I assume this was done so that adding a class to the button will then use that style. But, wouldn't it make more sense to call Initiate *after* the classes change instead of always creating every possible style on every control?

**Comment by grokys** [comment]
Hey, yes that would indeed make more sense, but I'm afraid it's an artifact of how our styling system currently works. I hope to be making some performance improvements to the styling system in the near future and this is definitely something I'd like to fix.

I think this issue is related: https://github.com/AvaloniaUI/Avalonia/issues/5027

**Reply by RationalFragile** [reply]
Thank you.
What I ended up doing is remaking my custom theming classes so that I don't use Avalonia's "Classes" like CSS Classes (because they really aren't the same ^^'). Instead, I make more individual UserControls that encapsulate those styles.

For example, instead of having `<Button Classes="Fancy" .../>` and Fancy Styles added to the App root, I have a `<FancyButton><FancyButton.Styles/>...</FancyButton>`.

Overall, I'm still extremely new to Avalonia (having never used WPF), and the more I use it the more I find better ways of achieving the same thing :) Thanks again!
